core/playlistmanager.go

```diff
@@ -0,0 +1,262 @@
+package core
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ericxtang/m3u8"
+	"github.com/golang/glog"
+	"github.com/livepeer/go-livepeer/drivers"
+	"github.com/livepeer/go-livepeer/net"
+	ffmpeg "github.com/livepeer/lpms/ffmpeg"
+	"github.com/livepeer/lpms/stream"
+)
+
+const LIVE_LIST_LENGTH uint = 10
+
+// XXX what if list will be longer?
+const FULL_LIST_LENGTH uint = 131072
+
+/*
+	PlaylistManager manages playlists and data for one video stream, backed by one object storage.
+*/
+type PlaylistManager interface {
+	ManifestID() ManifestID
+	// Implicitly creates master and media playlists
+	InsertHLSSegment(streamID StreamID, seg *stream.HLSSegment) (*net.TypedURI, error)
+	// InsertLink link to segment that already in the storage
+	// (Transcoder wrote segment in B's owned storage and returned link, B adds this link to video source)
+	// returns true if it was acutally inserted
+	InsertLink(turi *net.TypedURI, duration float64) (bool, error)
+
+	GetHLSMasterPlaylist() *m3u8.MasterPlaylist
+
+	GetHLSMediaPlaylist(streamID StreamID) *m3u8.MediaPlaylist
+
+	GetSession() drivers.OSSession
+	// Ends object storage's session
+	EndSession()
+}
+
+type BasicPlaylistManager struct {
+	storageSession drivers.OSSession
+	manifestID     ManifestID
+	// Live playlist used for broadcasting
+	masterLivePList *m3u8.MasterPlaylist
+	// Full playlist is stored into OS for later reference
+	masterFullPList *m3u8.MasterPlaylist
+	mediaLiveLists  map[StreamID]*m3u8.MediaPlaylist
+	mediaFullLists  map[StreamID]*m3u8.MediaPlaylist
+	mapSync         *sync.Mutex
+}
+
+// NewBasicPlaylistManager create new BasicPlaylistManager struct
+func NewBasicPlaylistManager(manifestID ManifestID,
+	storageSession drivers.OSSession) *BasicPlaylistManager {
+
+	bplm := &BasicPlaylistManager{
+		storageSession:  storageSession,
+		manifestID:      manifestID,
+		masterLivePList: m3u8.NewMasterPlaylist(),
+		masterFullPList: m3u8.NewMasterPlaylist(),
+		mediaLiveLists:  make(map[StreamID]*m3u8.MediaPlaylist),
+		mediaFullLists:  make(map[StreamID]*m3u8.MediaPlaylist),
+		mapSync:         &sync.Mutex{},
+	}
+	return bplm
+}
+
+func (mgr *BasicPlaylistManager) ManifestID() ManifestID {
+	return mgr.manifestID
+}
+
+func (mgr *BasicPlaylistManager) GetSession() drivers.OSSession {
+	return mgr.storageSession
+}
+
+// InsertHLSSegment ...
+func (mgr *BasicPlaylistManager) InsertHLSSegment(streamID StreamID,
+	seg *stream.HLSSegment) (*net.TypedURI, error) {
+
+	if err := mgr.isRightStream(streamID); err != nil {
+		return nil, err
+	}
+	sid := string(streamID)
+	fmpl := mgr.getFullPL(streamID)
+	if fmpl == nil {
+		panic("no full media pl")
+	}
+	if mgr.hasSegment(fmpl, seg.SeqNo) {
```

Comment:
Yes, because we can call `InsertLink` first, and `InsertHLSSegment` with the same segment after that. This happens when O responses with link to our own S3 storage (so we insert it), but we still want to get data to put it to another storage (local one currently, to save to disk), so we download data and calling `InsertHLSSegment`
> this is a sequential scan.
Forgot to optimize that, simple compare of last segment's SeqNo is enough here (if we assume that we're getting segments in order, if not, we can compare last 10)

Reply:
Shouldn't each backing store have different instances of `BasicPlaylistManager` ? When would a S3-backed playlist share data with a local storage playlist?

Reply:
They are. But they are hidden inside `CombinedPlaylistManager`

Reply:
When would we want to call both `InsertHLSSegment` and `InsertLink` for one playlist @darkdarkdragon ? Also, if these two functions store files in two different OS's, why do we need to check whether the segments might have been inserted twice, since they're in different places?

Reply:
The overall setup feels odd for a few reasons.

* We're potentially inserting it twice, which is a bit of a smell
* The 'correct' manifest management API to use is being dictated by the underlying object store.

Do you think we can unify the behavior a bit? Some thoughts:
* We always need a URL to insert into the manifest
* We always receive URLs from the transcoder
* Some URLs will belong to the same OS that's backing the manifest (eg, B's own S3)
  * In this case, insert the URL directly
* Download the URL indicated by the link
* Store the data from the download and retrieve the new URL
* Insert the new URL in the manifest

```
func (pm *BasicPlaylistManager) InsertSegment(streamId, url) {
  if pm.Store.Host != url.Host {
    data = Download(url)
    url = pm.Store.Put(data)
  }
  return pm.InsertManifest(streamId, url)
}
```

Reply:
I didn't go this way for couple of reasons:
1. For local B's storage, we don't have any urls - we already have binary blob
2. CombindePlaiylistManager can have any number of BasicPlaylistManagers inside, and potentially more than one of them will be downloading same data

Reply:
Case 1 -- we always get a URL from the transcoder. For local storage we can have blob API but there won't be a question about which one to use.

Case 2 is something we can optimize. It's one of the reasons I've been suggesting we hold off on concurrent OS until later -- combined playlists are a feature above and beyond what we currently require from OS.

Reply:
> but there won't be a question about which one to use.

Idea was to hide this behind interface, so code that uses it shouldn't choose anything

> is something we can optimize. 

Not sure how. And the reason I want it here - so we think about interfaces now, and not change them later.

Reply:
> Idea was to hide this behind interface, so code that uses it shouldn't choose anything

The code is choosing right now; it first attempts `InsertLink` then `InsertHLSSegment` if that fails. As mentioned, this has knock-on effects which makes the code odd. That's a sign the interface (or its semantics) could be improved.

> Not sure how. And the reason I want it here - so we think about interfaces now

It's certainly useful to think about how it could be implemented in order to design the interface. That doesn't mean we need to implement it, or have perfect interfaces up front.

For the combined playlist example, we could simply insert once for each distinct storage type, then maintain a mapping of storage types to playlists:

```
func NewCombinedPlaylist(playlists []PlaylistManager) *CombinedPlaylist {
  // create a mapping of storage backends and playlists, eg map<Storage, []Playlist>
  storageMap = CreateMap(playlists)
}
func CreateMap(playlists []PlaylistManager) map<Storage, []Playlist> {
  storageMap = make(map<Storage, []Playlist>)
  for p in playlists {
    key = p.Storage
    if !storageMap.HasKey(key) {
      storageMap[key] = {}
    }
    storageMap[key] = append(storageMap[key], p)
  }
}

func InsertBlob(streamID, blob) {
  for storage, playlists in storageMap {
    url = storage.Put(streamID, blob)
    for p in playlists {
      p.InsertManifest(streamID, url) // only inserts link; doesn't touch storage
    }
  }
}
func InsertLink(streamID, url) {
  ... repeat a variation of the above ...
}
```

Again, I just want to reiterate that we should set aside combined playlists for now until we're able to better design, build and test support for concurrent OS. Let's work on getting the core OS features in first.